web/src/echo.ts

```diff
@@ -376,14 +461,23 @@ export function update_message_lists({old_id, new_id}) {
     }
 }
 
-export function process_from_server(messages) {
-    const msgs_to_rerender = [];
-    const non_echo_messages = [];
+export function process_from_server(messages: ServerMessage[]): ServerMessage[] {
+    const msgs_to_rerender: Message[] = [];
+    const non_echo_messages: ServerMessage[] = [];
 
     for (const message of messages) {
         // In case we get the sent message before we get the send ACK, reify here
 
         const local_id = message.local_id;
+
+        if (local_id === undefined) {
+            // For messages that weren't locally echoed, we go through
+            // the "main" codepath that doesn't have to id reconciliation.
+            // We simply return non-echo messages to our caller.
```

Comment:
Should we add same comment twice?

Reply:
I believe the comments are "private" to their enclosing `if` blocks, hence I put them their twice.

Reply:
I ended up reorganizing the comments a bit for more clarity.